### DISCUSSION TITLE
Performance bottlenecked at threads=2

### DISCUSSION THREAD

**Discussion by lowfi117** [duckdb/duckdb#6017] [General]
I have a set of pretty large queries running on a 40K rows 20+ columns dataset. I can see performance boost when changing threads from 1 to 2, timing improved from 0.7x sec to 0.4x sec. But when I start to increase threads from 2 -> 24, I didn't see any improvement although I have a 12 cores cpu on my local machine. From [the blog post](https://duckdb.org/2021/05/14/sql-on-pandas.html), it seems suggest further improvement in performance with increasing of number of threads.
> Note that while DuckDB can scale far beyond two threads, Google Colab only supports two.

Do I miss anything?

Setup of my environment:
1. Use duckdb Python API. Change threads by `duck_cur.execute(f"PRAGMA threads={duckdb_nthreads}")`
2. Ubuntu environment.
3. 12 cores CPU, 64 RAM.
4. The query contains creating a temp table and histogram computation on the columns.

**Comment by Mytherin** [comment]
DuckDB partitions data in sets of 100K rows by default - if you are doing expensive calculations over very small tables then the current model of parallelism won't work as well. You could try enabling smaller grained parallelism using `PRAGMA verify_parallelism` and seeing if that helps (although this is really more intended for testing than practical use).

**Comment by lowfi117** [comment]
Thanks for the helpful hint. My bad I miscount the rows, it's actually 190K. So it perfect explains why the performance plateau at 2 threads.

I experiment with `PRAGMA verify_parallelism`, it shows the performance is sort of optimal at 4 (0.3x secs with some variance) compare to 1 (0.7x sec), 2 (0.4x sec) and 8 (0.4 sec). Yeah, so small table expensive calculation would be the issue anyway.

But my larger goal is to have support quick queries on table with 50x-100x more rows than currently I have through vertical scaling (48+ cores CPU on k8s pod). So it's large data and expensive computation at same time. Great to hear that the parallelism in DuckDB would be more effective when dealing with 50x-100x more data. Does my use case sounds right to you?

**Reply by Mytherin** [reply]
Yes, DuckDB is optimized for efficiently processing many rows on a single box. It should definitely scale better with more rows.
